### DISCUSSION TITLE
Cape Cod roll-forward

### DISCUSSION THREAD

**Discussion by jbogaardt** [casact/chainladder-python#406] [General]
chainladder-python community,

I have a question and looking to solicit thoughts on the functionality of the `CapeCod` method.  It's pretty well known that the generalized `CapeCod` is really just a `BornhuetterFerguson` method with a bit more sophisticated, and triangle-centric algorithm for picking the apriori expectation. The apriori algorithm draws from existing origin period triangle data to select apriori ultimates for the same origin periods.  My question is whether it is even feasible to roll-forward the `CapeCod` for a new diagonal and new origin period?

Rolling forward is typically done with the `predict` method.  The current implementation of the `CapeCod` recomputes the apriori on the triangle being predicted, but this just feels like refitting to me (albeit holding the development patterns constant).

Perhaps more general than the `CapeCod`, should it be feasible to roll forward ANY estimator where the fitted parameters are tied to the specific origin periods of the prior triangle. This is a non-issue for `Chainladder` and `BornhuetterFerguson` because these methods have parameters `ldf_` and `apriori` that are independent of the origin being used.

**Comment by kennethshsu** [comment]
It's quiet, I'll reply with a question.

> This is a non-issue for `Chainladder` and `BornhuetterFerguson` because these methods have parameters `ldf_` and `apriori` that are independent of the origin being used.

I don't think I follow your thought on this. When you are rolling forward, you are basically using say, 4/30 data for analysis, but you need to close the book for 6/30? Essentially you are trying to roll forward May and June's ultimates? When you are rolling forward this way, don't you just pick an apriori expectation * your exposure? What does that have anything to do with `Chainladder` or `BornhuetterFerguson`?

**Comment by mlbugs** [comment]
He’s talking about fitting cape cod to one triangle and then applying the fitted parameters to a different triangle, no?On Dec 15, 2022, at 6:31 PM, Kenneth S. Hsu ***@***.***> wrote:﻿
It's quiet, I'll reply with a question.

This is a non-issue for Chainladder and BornhuetterFerguson because these methods have parameters ldf_ and apriori that are independent of the origin being used.

I don't think I follow your thought on this. When you are rolling forward, you are basically using say, 4/30 data for analysis, but you need to close the book for 6/30? Essentially you are trying to roll forward May and June's ultimates? When you are rolling forward this way, don't you just pick an apriori expectation * your exposure? What does that have anything to do with Chainladder or BornhuetterFerguson?

**Comment by jbogaardt** [comment]
Yes, fitting to one triangle and predicting on a new one.

Suppose I fit a `CapeCod` to a triangle with origins 2011-2021. I want to use that model to predict 2012-2022, but the `CapeCod` apriori is specific to the origins of the original triangle. I don't think I can just use the old apriori because any applied trend/onlevel is specific to 2011-2021. It seems I'd need to fit a new model for a new period and the `CapeCod` should error when attempting to predict on different origins, right?

**Comment by kennethshsu** [comment]
Ahh got it. Yes you are right, unless the apriori is adjusted for trend and further on-leveling.

However, the model should work though if there is no trend or adjustment made in the original triangle I think. I don't think it makes sense to error it out though, it should be up to the user to notice that further adjustment is needed. This isn't the model's fault, but an improper use of `CapeCod`. Do you agree?

**Reply by jbogaardt** [reply]
That makes sense. I could see cases where you may have a regional apriori, but want to apply it down to a specific geographic state. So long as the origin lines up between the fitted regional model and the predicted state level model, it should be permitted. Perhaps we only error if there are new origins in the triangle to be predicted that don't also exist in the fitted model.
